pkg/blobstore/cassandra/cassandra_blob_access.go

```diff
@@ -0,0 +1,673 @@
+package cassandra
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"math"
+	"math/rand"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/buildbarn/bb-storage/pkg/blobstore"
+	"github.com/buildbarn/bb-storage/pkg/blobstore/buffer"
+	"github.com/buildbarn/bb-storage/pkg/blobstore/slicing"
+	"github.com/buildbarn/bb-storage/pkg/capabilities"
+	bbdigest "github.com/buildbarn/bb-storage/pkg/digest"
+	"github.com/buildbarn/bb-storage/pkg/util"
+
+	remoteexecution "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/gocql/gocql"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	writeConsistency            = gocql.LocalQuorum
+	readConsistency             = gocql.LocalQuorum
+	lastAccessUpdateConsistency = gocql.LocalOne
+
+	simultaneousWorkers = 100
+
+	// The number of workers to use for updating the `last_access` field
+	lastAccessUpdateWorkerCount = 50
+)
+
+var (
+	cassandraBlobStorePrometheusMetrics sync.Once
+
+	segmentWriteHist = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "buildbarn",
+		Subsystem: "storage",
+		Name:      "cassandra_segment_write_duration_in_seconds",
+		Help:      "Amount of time in seconds it takes to write each segment to Cassandra, measured in seconds",
+		Buckets:   util.DecimalExponentialBuckets(-3, 6, 2),
+	})
+
+	segmentReadHist = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "buildbarn",
+		Subsystem: "storage",
+		Name:      "cassandra_segment_read_duration_in_seconds",
+		Help:      "Amount of time in seconds it takes to read each segment from Cassandra",
+		Buckets:   util.DecimalExponentialBuckets(-3, 6, 2),
+	})
+
+	totalWritesHist = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "buildbarn",
+		Subsystem: "storage",
+		Name:      "cassandra_total_write_duration_in_seconds",
+		Help:      "Total amount of time in seconds it takes to write an entire bytestream to Cassandra",
+		Buckets:   util.DecimalExponentialBuckets(-3, 6, 2),
+	})
+
+	lastAccessUpdatesFailures = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "buildbarn",
+		Subsystem: "storage",
+		Name:      "cassandra_last_access_failure_count",
+		Help:      "Count of the number of times the `last_access` field could not be updated",
+	})
+)
+
+type job struct {
+	task func()
+}
```

Comment:
Do we expect that `job` ever gets additional members? If not, probably remove it and use the callback type directly?